### PR TITLE
grafana/ui: Enable React based options editors for Datasource plugins

### PR DESCRIPTION
**Pull request by dprokop**
Side effect of #16731 discovery phase. Follows the same convention (`setEditor`) as React panels do.

### CODE REVIEW

File: public/app/features/datasources/settings/DataSourceSettingsPage.tsx [hunk 1]
```diff
@@ -36,10 +37,12 @@ export interface Props {
   setDataSourceName: typeof setDataSourceName;
   updateDataSource: typeof updateDataSource;
   setIsDefault: typeof setIsDefault;
+  plugin?: DataSourcePlugin<any, any>;
```
Comment (torkelo) [inline public/app/features/datasources/settings/DataSourceSettingsPage.tsx]: you do not need the any, any the types are optional?

Comment (dprokop) [reply public/app/features/datasources/settings/DataSourceSettingsPage.tsx]: You are right, but Options are not optional. Gonna make options default to be {}

File: packages/grafana-ui/src/types/datasource.ts [hunk 2]
```diff
@@ -4,15 +4,24 @@ import { PluginMeta } from './plugin';
...
   DataSourceClass: DataSourceConstructor<TQuery>;
   components: DataSourcePluginComponents<TQuery>;
+  editor?: React.ComponentType<DataSourcePluginOptionsEditorProps<TOptions>>;
 
   constructor(DataSourceClass: DataSourceConstructor<TQuery>) {
     this.DataSourceClass = DataSourceClass;
     this.components = {};
   }
 
+  setEditor(editor: React.ComponentType<DataSourcePluginOptionsEditorProps<TOptions>>) {
```
Comment (torkelo) [inline packages/grafana-ui/src/types/datasource.ts]: Editor is a bit too non-descriptive for a data source, as it already as a QueryEditor. 

Think setConfigEditor is more descriptive (as the current angular version is called ConfigCtrl). Also think this should be on the components sub object along with all the other components.

**Comment by ryantxu** [comment]
Check #16586 also... It includes something like this

**Comment by dprokop** [comment]
@ryantxu yeah, this one is waaaaay simpler. But i see a future in your `PluginWithConfig` interface ;)

**Pull request merged**